cmd/DownloadFFISSpreadsheet/handler.go

```diff
@@ -0,0 +1,105 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"github.com/aws/aws-lambda-go/events"
+	s3manager "github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/usdigitalresponse/grants-ingest/internal/log"
+	"github.com/usdigitalresponse/grants-ingest/pkg/grantsSchemas/ffis"
+)
+
+// error constants
+var (
+	ErrDownloadFailed = fmt.Errorf("error downloading file")
+)
+
+type S3UploaderAPI interface {
+	Upload(ctx context.Context,
+		params *s3.PutObjectInput,
+		optFns ...func(*s3manager.Uploader)) (*s3manager.UploadOutput, error)
+}
+
+type HTTPClientAPI interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+func handleSQSEvent(ctx context.Context, sqsEvent events.SQSEvent, s3Uploader S3UploaderAPI, httpClient HTTPClientAPI) error {
+	msg := sqsEvent.Records[0].Body
+	log.Info(logger, "Received message", "message", msg)
+	var ffisMessage ffis.FFISMessageDownload
+	err := json.Unmarshal([]byte(msg), &ffisMessage)
+	if err != nil {
+		return fmt.Errorf("error unmarshalling SQS message: %w", err)
+	}
+	fileStream, err := downloadFile(ctx, ffisMessage, httpClient)
+	if err != nil {
+		return fmt.Errorf("error parsing SQS message: %w", err)
+	}
+	defer fileStream.Close()
+	err = writeToS3(ctx, s3Uploader, fileStream, ffisMessage.SourceFileKey)
+	return err
+}
+
+func downloadFile(ctx context.Context, msg ffis.FFISMessageDownload, httpClient HTTPClientAPI) (stream io.ReadCloser, err error) {
+	resp, err := startDownload(ctx, httpClient, msg.DownloadURL)
+	if err != nil {
+		return nil, fmt.Errorf("error downloading file: %w", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("error downloading file: %w", ErrDownloadFailed)
+	}
+	log.Debug(logger, "Downloaded file", "url", msg.DownloadURL)
+	return resp.Body, nil
+}
+
+// startDownload starts a new download request and returns the response.
+// Failed requests retry until env.MaxDownloadBackoff elapses.
+// Returns a non-nil error if the request either could not be initialized or never succeeded.
+func startDownload(ctx context.Context, httpClient HTTPClientAPI, url string) (resp *http.Response, err error) {
```

Comment:
@TylerHendrickson This mostly a shameful copy paste from https://github.com/usdigitalresponse/grants-ingest/blob/4227c1cf0a13127229adac1b676623b20838f1e4/cmd/DownloadGrantsGovDB/handler.go#L78... Want me to extract this into an `internal` package or anything like that?

Reply:
Yeah, I had a similar thought about extracting this into `internal`. Given that this branch is working well, I'd hesitate to make that a blocker for this particular issue, but I think it would make sense to create a new issue that captures the work needed to implement the helper library and refactor the `DownloadGrantsGovDB` and `DownloadFFISSpreadsheet` Lambda functions to make use of it.

Feel free to create the issue if you'd like, otherwise I'll try to get to that a bit later today.